client/my-sites/stats/stats-notices/index.tsx

```diff
@@ -13,116 +14,83 @@ import hasSiteProductJetpackStatsFree from 'calypso/state/sites/selectors/has-si
-		isWpcom &&
-		! isVip &&
-		! isP2 &&
-		! isOwnedByTeam51;
-	// Show the notice if the site is Jetpack or it is Odyssey Stats.
-	const showUpgradeNoticeOnOdyssey = config.isEnabled( 'stats/paid-stats' ) && isOdysseyStats;
-	const showUpgradeNoticeForJetpackNotAtomic =
-		config.isEnabled( 'stats/paid-stats' ) && isSiteJetpackNotAtomic;
+		isOdysseyStats,
+		isWpcom,
+		isVip,
+		isP2,
+		isOwnedByTeam51,
+		hasPaidStats,
+		hasFreeStats,
+		isSiteJetpackNotAtomic,
+		statsPurchaseSuccess,
+	};
 
-	const showDoYouLoveJetpackStatsNotice =
-		isDoYouLoveJetpackStatsVisible &&
-		( showUpgradeNoticeOnOdyssey ||
-			showUpgradeNoticeForJetpackNotAtomic ||
-			showUpgradeNoticeForWpcomSites ) &&
-		// Show the notice if the site has not purchased the paid stats product.
-		! hasPaidStats &&
-		hasLoadedPurchases;
+	const {
+		isLoading,
+		isError,
+		data: serverNoticesVisibility,
+	} = useNoticesVisibilityQueryRaw( siteId );
 
-	return (
-		<>
-			{ showDoYouLoveJetpackStatsNotice && (
-				<DoYouLoveJetpackStatsNotice siteId={ siteId } hasFreeStats={ hasFreeStats } />
-			) }
-		</>
-	);
-};
-
-const PostPurchaseNotices = ( {
-	siteId,
-	statsPurchaseSuccess,
-	isOdysseyStats,
-}: StatsNoticesProps ) => {
-	// Check if the GET param is passed to show the Free or Paid plan purchase notices
-	const showFreePlanPurchaseSuccessNotice = statsPurchaseSuccess === 'free';
-	const showPaidPlanPurchaseSuccessNotice = statsPurchaseSuccess === 'paid';
+	const { hasLoadedPurchases } = usePurchasesToUpdateSiteProducts( isOdysseyStats, siteId );
 
-	const [ paramRemoved, setParamRemoved ] = useState( false );
+	if ( ! hasLoadedPurchases || isLoading || isError ) {
+		return null;
+	}
 
-	const removeParam = () => {
-		if ( ! statsPurchaseSuccess || paramRemoved ) {
-			return;
-		}
-		// Ensure it runs only once.
-		setParamRemoved( true );
-		const newUrlObj = removeStatsPurchaseSuccessParam( window.location.href, !! isOdysseyStats );
-		// Odyssey would try to hack the URL on load to remove duplicate params. We need to wait for that to finish.
-		setTimeout( () => {
-			window.history.replaceState( null, '', newUrlObj.toString() );
-			if ( isOdysseyStats ) {
-				// We need to update the page base if it changed. Otherwise, pagejs won't be able to find the routes.
-				page.base( `${ newUrlObj.pathname }${ newUrlObj.search }` );
-			}
-		}, 300 );
-	};
+	const calculatedNoticesVisibility = ensureOnlyOneNoticeVisible(
+		serverNoticesVisibility,
+		noticeOptions
+	);
 
-	return (
-		<>
-			{ /* TODO: Consider combining/refactoring these components into a single component */ }
-			{ showPaidPlanPurchaseSuccessNotice && (
-				<PaidPlanPurchaseSuccessJetpackStatsNotice onNoticeViewed={ removeParam } />
-			) }
-			{ showFreePlanPurchaseSuccessNotice && (
-				<FreePlanPurchaseSuccessJetpackStatsNotice
-					siteId={ siteId }
-					onNoticeViewed={ removeParam }
-				/>
-			) }
-		</>
+	return ALL_STATS_NOTICES.map(
```

Comment:
The components array seems not to be accepted as the return type of `JSX element`. Could we find out the only notice component and return it?
```suggestion
	const onlyOneNotice = ALL_STATS_NOTICES.find(
		( notice ) => calculatedNoticesVisibility[ notice.noticeId ]
	);

	return onlyOneNotice ? <onlyOneNotice.component { ...noticeOptions } /> : null;
```

<img width="1205" alt="截圖 2023-08-24 上午1 21 42" src="https://github.com/Automattic/wp-calypso/assets/6869813/1320530f-5a1a-4370-b0c7-4424c3f87c19">

Reply:
Should be fixed in 11d3344bad; weirdly I am not able to see these errors 🤔